### PR TITLE
support kubectl get --raw /api/v1/nodes/{node}/proxy/metrics

### PR DESCRIPTION
**Pull request by Abirdcfly**
**What type of PR is this?**

 /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubeedge/kubeedge/issues/2436

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cloud/pkg/cloudstream/tunnelserver.go [hunk 1]
```diff
@@ -39,13 +39,15 @@ type TunnelServer struct {
 	container *restful.Container
 	upgrader  websocket.Upgrader
 	sync.Mutex
-	sessions map[string]*Session
+	sessions   map[string]*Session
+	nodeNameIP map[string]string
```
Comment (daixiang0) [inline cloud/pkg/cloudstream/tunnelserver.go:43]: Could use sync.Map here?

Comment (Abirdcfly) [reply cloud/pkg/cloudstream/tunnelserver.go]: update to use sync.Map

**Review by chendave — COMMENTED** [review]
I would hope that eventually there is testcases provided for each functionality that is being added.

File: cloud/pkg/cloudstream/tunnelserver.go [hunk 2]
```diff
@@ -79,6 +80,18 @@ func (s *TunnelServer) getSession(id string) (*Session, bool) {
...
+func (s *TunnelServer) addNodeIP(node, ip string) {
+	s.nodeNameIP.Store(node, ip)
+}
+
+func (s *TunnelServer) getNodeIP(node string) (string, bool) {
+	ip, ok := s.nodeNameIP.Load(node)
+	if !ok {
+		return "", ok
+	}
+	return ip.(string), ok
```
Comment (chendave) [inline cloud/pkg/cloudstream/tunnelserver.go:92]: those check is redundant, you have checked the bool in the place where it is got called.

Comment (Abirdcfly) [reply cloud/pkg/cloudstream/tunnelserver.go:92]: if not check，ip.(string) will get a panic.

**Comment by Abirdcfly** [comment]
> I would hope that eventually there is testcases provided for each functionality that is being added.

there is no testcase for feature "support kubectl log / kubectl metric / kubectl exec". so no testcase for this small improvement. I hope I can add some testcase for that feature. but may be another pull request.

**Review by fisherxu — COMMENTED** [review]
Thanks for the fixinig.

File: cloud/pkg/cloudstream/streamserver.go [hunk 3]
```diff
@@ -153,6 +154,14 @@ func (s *StreamServer) getMetrics(r *restful.Request, w *restful.Response) {
 	}()
 
 	sessionKey := strings.Split(r.Request.Host, ":")[0]
+	if forwardedURI := r.Request.Header.Get("X-Forwarded-Uri"); forwardedURI != "" {
+		if t := strings.Split(forwardedURI, "/"); strings.HasPrefix(forwardedURI, "/api/v1/nodes/") && len(t) > 6 {
+			sessionKey = t[4]
+			if ip, ok := s.tunnel.getNodeIP(sessionKey); ok {
+				r.Request.Host = fmt.Sprintf("%s:%d", ip, constants.ServerPort)
+			}
+		}
```
Comment (fisherxu) [inline cloud/pkg/cloudstream/streamserver.go:163]: For this request, the node ip doesn't exist in `r.Request.Host`?

Comment (Abirdcfly) [reply cloud/pkg/cloudstream/streamserver.go:163]: https://github.com/kubeedge/kubeedge/issues/2436#issuecomment-754580697 I add some explanation.😂

Comment (fisherxu) [reply cloud/pkg/cloudstream/streamserver.go:163]: Looks good, thanks for the fixing.

**Pull request merged**